flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunnerImpl.java

```diff
@@ -0,0 +1,445 @@
+import org.apache.flink.runtime.executiongraph.ArchivedExecutionGraph;
+import org.apache.flink.runtime.highavailability.HighAvailabilityServices;
+import org.apache.flink.runtime.highavailability.RunningJobsRegistry;
+import org.apache.flink.runtime.highavailability.RunningJobsRegistry.JobSchedulingStatus;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.jobmanager.OnCompletionActions;
+import org.apache.flink.runtime.jobmaster.factories.JobMasterServiceFactory;
+import org.apache.flink.runtime.leaderelection.LeaderContender;
+import org.apache.flink.runtime.leaderelection.LeaderElectionService;
+import org.apache.flink.runtime.messages.Acknowledge;
+import org.apache.flink.runtime.rpc.FatalErrorHandler;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.FlinkException;
+import org.apache.flink.util.function.FunctionUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+
+import java.io.IOException;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * The runner for the job manager. It deals with job level leader election and make underlying job manager
+ * properly reacted.
+ */
+public class JobManagerRunnerImpl implements LeaderContender, OnCompletionActions, JobManagerRunner {
+
+	private static final Logger log = LoggerFactory.getLogger(JobManagerRunnerImpl.class);
+
+	// ------------------------------------------------------------------------
+
+	/** Lock to ensure that this runner can deal with leader election event and job completion notifies simultaneously. */
+	private final Object lock = new Object();
+
+	/** The job graph needs to run. */
+	private final JobGraph jobGraph;
+
+	/** Used to check whether a job needs to be run. */
+	private final RunningJobsRegistry runningJobsRegistry;
+
+	/** Leader election for this job. */
+	private final LeaderElectionService leaderElectionService;
+
+	private final LibraryCacheManager libraryCacheManager;
+
+	private final Executor executor;
+
+	private final JobMasterService jobMasterService;
+
+	private final FatalErrorHandler fatalErrorHandler;
+
+	private final CompletableFuture<ArchivedExecutionGraph> resultFuture;
+
+	private final CompletableFuture<Void> terminationFuture;
+
+	private CompletableFuture<Void> leadershipOperation;
+
+	/** flag marking the runner as shut down. */
+	private volatile boolean shutdown;
+
+	private volatile CompletableFuture<JobMasterGateway> leaderGatewayFuture;
+
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Exceptions that occur while creating the JobManager or JobManagerRunner are directly
```

Comment:
Since legacy `JobManager` has been removed, it can be consider how we deal with naming issue of `JobMaster` vs `JobManager`. Many times we refer to `JobManager`/JM which actually means `ClusterEntrypoint`.

Reply:
True, we are currently not very consistent with our terminology and should improve. I think, however, that this is out of scope of this PR.